aiohomekit/controller/coap/connection.py

```diff
@@ -0,0 +1,1097 @@
+        return None
+
+    def to_dict(self):
+        return {
+            "aid": self.instance_id,
+            "services": [service.to_dict() for service in self.services],
+        }
+
+
+@dataclass
+class Pdu09AccessoryContainer(TLVStruct):
+    accessory: Pdu09Accessory = tlv_entry(HAP_TLV.kTLVHAPParamUnknown_19)
+
+
+@dataclass
+class Pdu09Database(TLVStruct):
+    _accessories: Sequence[Pdu09AccessoryContainer] = tlv_entry(
+        HAP_TLV.kTLVHAPParamUnknown_18
+    )
+
+    @property
+    def accessories(self):
+        return [container.accessory for container in self._accessories]
+
+    def find_characteristic_by_aid_iid(self, aid, iid):
+        for accessory in self.accessories:
+            if accessory.instance_id == aid:
+                return accessory.find_characteristic_by_iid(iid)
+        return None
+
+    # return first matching iid
+    def find_characteristic_by_iid(self, iid):
+        for accessory in self.accessories:
+            characteristic = accessory.find_characteristic_by_iid(iid)
+            if characteristic:
+                return characteristic
+        return None
+
+    def to_dict(self):
+        return [accessory.to_dict() for accessory in self.accessories]
+
+
+class EncryptionContext:
+    def __init__(self, recvCtx, sendCtx, eventCtx, uri, coapCtx):
+        self.recvCtr = 0
+        self.recvCtx = recvCtx
+        self.sendCtr = 0
+        self.sendCtx = sendCtx
+        self.eventCtr = 0
+        self.eventCtx = eventCtx
+
+        self.coapCtx = coapCtx
+        self.lock = asyncio.Lock()
+        self.uri = uri
+
+    def decrypt(self, enc_data):
+        logger.debug("DECRYPT counter=%d" % (self.recvCtr,))
+        dec_data = self.recvCtx.decrypt(
+            struct.pack("=4xQ", self.recvCtr), enc_data, b""
+        )
+        self.recvCtr += 1
+        return dec_data
+
+    def decrypt_event(self, enc_data):
+        dec_data = self.eventCtx.decrypt(
+            struct.pack("=4xQ", self.eventCtr), enc_data, b""
+        )
+        self.eventCtr += 1
+        return dec_data
+
+    def encrypt(self, dec_data):
+        logger.debug("ENCRYPT counter=%d" % (self.sendCtr,))
+        enc_data = self.sendCtx.encrypt(
+            struct.pack("=4xQ", self.sendCtr), dec_data, b""
+        )
+        self.sendCtr += 1
+        return enc_data
+
+    async def post(self, payload, timeout=16.0):
+        async with self.lock:
+            payload = self.encrypt(payload)
+
+            try:
+                request = Message(code=Code.POST, payload=payload, uri=self.uri)
+                response = await asyncio.wait_for(
+                    self.coapCtx.request(request).response, timeout=timeout
+                )
+            except (NetworkError, asyncio.TimeoutError):
+                raise AccessoryDisconnectedError("Request timeout")
+
+            if response.code != Code.CHANGED:
+                logger.warn(f"CoAP POST returned unexpected code {response}")
+
+            try:
```

Comment:
Theres no i/o in the decrypt retry code. So can we pull it out of `post` so thats its more testable, and add some tests for it.